awx/ui/client/features/credentials/add-edit-credentials.controller.js

```diff
@@ -0,0 +1,610 @@
+        const metaIsActive = name === 'metadata';
+        vm.inputSources.tabs.credential._active = !metaIsActive;
+        vm.inputSources.tabs.credential._disabled = false;
+        vm.inputSources.tabs.metadata._active = metaIsActive;
+        vm.inputSources.tabs.metadata._disabled = false;
+    }
+
+    function unsetInputSourceTabs () {
+        vm.inputSources.tabs.credential._active = false;
+        vm.inputSources.tabs.credential._disabled = false;
+        vm.inputSources.tabs.metadata._active = false;
+        vm.inputSources.tabs.metadata._disabled = false;
+    }
+
+    vm.onInputSourceClear = (field) => {
+        vm.form[field].tagMode = true;
+        vm.form[field].asTag = false;
+        vm.form[field]._value = '';
+        vm.form[field]._tagValue = '';
+        vm.inputSources.items = vm.inputSources.items
+            .filter(({ input_field_name }) => input_field_name !== field);
+        vm.inputSources.changedInputFields.push(field);
+    };
+
+    vm.onInputSourceOpen = (field) => {
+        // We get here when the input source lookup modal for a field is opened. If source
+        // credential and metadata values for this field already exist in the initial API data
+        // or from it being set during a prior visit to the lookup, we initialize the lookup with
+        // these values here before opening it.
+        const sourceItem = vm.inputSources.items
+            .find(({ input_field_name }) => input_field_name === field);
+        if (sourceItem) {
+            const { source_credential, summary_fields } = sourceItem;
+            const { source_credential: { credential_type_id, name } } = summary_fields;
+            vm.inputSources.credentialId = source_credential;
+            vm.inputSources.credentialName = name;
+            vm.inputSources.credentialTypeId = credential_type_id;
+            vm.inputSources._value = credential_type_id;
+        } else {
+            vm.inputSources.credentialId = null;
+            vm.inputSources.credentialName = null;
+            vm.inputSources.credentialTypeId = null;
+            vm.inputSources._value = null;
+        }
+
+        setInputSourceTab('credential');
+        vm.inputSources.field = field;
+    };
+
+    vm.onInputSourceClose = () => {
+        // We get here if the lookup was closed or canceled so we clear the state for the lookup
+        // and metadata form without storing any changes.
+        vm.inputSources.field = null;
+        vm.inputSources.credentialId = null;
+        vm.inputSources.credentialName = null;
+        vm.inputSources.metadataInputs = null;
+        unsetInputSourceTabs();
+    };
+
+    /**
+     * Extract the current set of input values from the metadata form and reshape them to a
+     * metadata object that can be sent to the api later or reloaded when re-opening the form.
+     */
+    function getMetadataFormSubmitData ({ inputs }) {
+        const metadata = Object.assign({}, ...inputs._group
```

Comment:
Instead of chaining `.filter`, `.map`, and `Object.assign` to craft the final form of the array you are processing, I think a `.reduce` method here fits the bill nicely because it will do all three in one go, and is meant for these types of conversions:

```function getMetadataFormSubmitData ({ inputs }) {
        return inputs._group
            .reduce((formatted, { id, _value }) => {
                if (_value) {
                    formatted[id] = _value;
                }
                return formatted;
            }, {});
        } ```

Reply:
👍 works for me